introspection.go

```diff
@@ -4,13 +4,80 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"net/http"
 
 	"github.com/vektah/gqlparser/v2/ast"
 )
 
+// IntrospectOptions represents the options for the IntrospectAPI function
+type IntrospectOptions struct {
+	wares  []NetworkMiddleware
+	client *http.Client
+	ctx    context.Context
+}
+
+// Context returns either a given context or an instance of the context.Background
+func (o *IntrospectOptions) Context() context.Context {
+	if o.ctx == nil {
+		return context.Background()
+	}
+	return o.ctx
+}
+
+// Apply applies the options to a given queryer
+func (o *IntrospectOptions) Apply(queryer Queryer) Queryer {
+	if q, ok := queryer.(QueryerWithMiddlewares); ok && len(o.wares) > 0 {
+		queryer = q.WithMiddlewares(o.wares)
+	}
+	if q, ok := queryer.(HTTPQueryer); ok && o.client != nil {
+		queryer = q.WithHTTPClient(o.client)
+	}
+	return queryer
+}
+
+func mergeIntrospectOptions(opts ...*IntrospectOptions) *IntrospectOptions {
+	res := &IntrospectOptions{}
+	for _, opt := range opts {
+		if len(opt.wares) > 0 {
+			res.wares = append(res.wares, opt.wares...)
+		}
+		if opt.client != nil {
+			res.client = opt.client
+		}
+		if opt.ctx != nil {
+			res.ctx = opt.ctx
+		}
+	}
+	return res
+}
+
+// IntrospectWithMiddlewares returns an instance of graphql.IntrospectOptions with given middlewares
+// to be pass to an instance of a graphql.Queryer by the IntrospectOptions.Apply function
+func IntrospectWithMiddlewares(wares ...NetworkMiddleware) *IntrospectOptions {
+	return &IntrospectOptions{
+		wares: wares,
+	}
+}
+
+// IntrospectWithHTTPClient returns an instance of graphql.IntrospectOptions with given client
+// to be pass to an instance of a graphql.Queryer by the IntrospectOptions.Apply function
+func IntrospectWithHTTPClient(client *http.Client) *IntrospectOptions {
```

Comment:
How would a user compose `IntrospectWithHTTPClient` and `IntrospectWithContext`? I think a better approach would be for `IntrospectOptions` to follow the same pattern as [gateway.Option](https://github.com/nautilus/gateway/blob/master/gateway.go#L223) which returns a function that mutates an initial state. This way a user could do something like `graphql.IntrospectRemoteSchema(graphql.IntrospectWithClient(...), graphql.IntrospectWithContext())`. Thoughts?

Reply:
I can do it, we need to modify the querier object, but not a context, since it returns a context that should be used in Query function. So the IntrospectWithContext will a custom case of the Options.

I use same implementation of the options, as go mongo driver.

it can be used as `IntrospectRemoteSchema("http://example.com/graphql", IntrospectWithMiddlewares(...), IntrospectWithClietn(...), IntrospectWithContext(...))`

Reply:
> How would a user compose IntrospectWithHTTPClient and IntrospectWithContext? 
there is a special function not merge the options `mergeIntrospectOptions`

Reply:
Ah, I see what you mean now. I missed the `merge` function that combines them all. I think this is good to go!